### PR TITLE
feat: add cases of rfc0031

### PR DESCRIPTION
**Pull request by keroro520**
```text
┌─────────────────────┬───────────────────────┬───────────────────────┐
│                     │                       │                       │
│ block.extension.len │  v2019                │  v2021                │
├─────────────────────┼───────────────────────┼───────────────────────┤
│ None                │  Ok                   │  Ok                   │
├─────────────────────┼───────────────────────┼───────────────────────┤
│ Some(0)             │  Err(UNKNOWN_FIELDS)  │  Err(EMPTY_EXT)       │
├─────────────────────┼───────────────────────┼───────────────────────┤
│ Some(1)             │  Err(UNKNOWN_FIELDS)  │  Ok                   │
├─────────────────────┼───────────────────────┼───────────────────────┤
│ Some(16)            │  Err(UNKNOWN_FIELDS)  │  Ok                   │
├─────────────────────┼───────────────────────┼───────────────────────┤
│ Some(32)            │  Err(UNKNOWN_FIELDS)  │  Ok                   │
├─────────────────────┼───────────────────────┼───────────────────────┤
│ Some(64)            │  Err(UNKNOWN_FIELDS)  │  Ok                   │
├─────────────────────┼───────────────────────┼───────────────────────┤
│ Some(96)            │  Err(UNKNOWN_FIELDS)  │  Ok                   │
├─────────────────────┼───────────────────────┼───────────────────────┤
│ Some(97)            │  Err(UNKNOWN_FIELDS)  │  Err(MAX_LIMIT)       │
└─────────────────────┴───────────────────────┴───────────────────────┘
```

### CODE REVIEW

File: ckb-integration-test/src/case/rfc0031/v2019.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            node_options: vec![NodeOptions {
+                node_name: String::from("node2021"),
+                ckb_binary: CKB2021.read().unwrap().clone(),
+                initial_database: "testdata/db/Epoch2V2TestData",
+                chain_spec: "testdata/spec/ckb2021",
+                app_config: "testdata/config/ckb2021",
+            }]
+            .into_iter()
+            .collect(),
+        }
```
Comment (zhangsoledad) [inline ckb-integration-test/src/case/rfc0031/v2019.rs:28]: should be 2019？

File: ckb-integration-test/src/case/rfc0031/v2019.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+        let cases = vec![
+            (node2021, None, Ok(())),
+            (node2021, Some(0), Err(ERROR_UNKNOWN_FIELDS)),
+            (node2021, Some(1), Err(ERROR_UNKNOWN_FIELDS)),
+            (node2021, Some(16), Err(ERROR_UNKNOWN_FIELDS)),
+            (node2021, Some(32), Err(ERROR_UNKNOWN_FIELDS)),
+            (node2021, Some(64), Err(ERROR_UNKNOWN_FIELDS)),
+            (node2021, Some(96), Err(ERROR_UNKNOWN_FIELDS)),
+            (node2021, Some(97), Err(ERROR_UNKNOWN_FIELDS)),
```
Comment (doitian) [inline ckb-integration-test/src/case/rfc0031/v2019.rs:43]: TODO: how to create a block rejected by `ckb_types`.

We need to support sending arbitrary binary via p2p protocols to peer for test.

File: ckb-integration-test/src/case/rfc0031/v2019.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+    fn run(&self, nodes: Nodes) {
+        let node2021 = nodes.get_node("node2021");
+        assert!(!is_rfc0224_switched(node2021));
+
+        let cases = vec![
+            (node2021, None, Ok(())),
+            (node2021, Some(0), Err(ERROR_UNKNOWN_FIELDS)),
```
Comment (doitian) [inline ckb-integration-test/src/case/rfc0031/v2019.rs:37]: We disallow extension which size is 0, can you double check with @yangby-cryptape ?

Comment (yangby-cryptape) [reply ckb-integration-test/src/case/rfc0031/v2019.rs:37]: > We disallow extension which size is 0, can you double check with @yangby-cryptape ?

Yes, we have discussed this when I developed this: empty extension is not allowed.

File: ckb-integration-test/src/case/rfc0031/v2019.rs (hunk 1 above)
Comment (keroro520) [reply ckb-integration-test/src/case/rfc0031/v2019.rs:28]: I will change the name back to `RFC0031BeforeSwitch`/`RFC0031AfterSwitch`.

**Pull request merged**